Issues with reading `application.yml` properties into @ConfigurationProperties bean containing `LocalDate`

### Expected Behavior

@ConfigurationProperties beans having properties of type `java.time.LocalDate` should be mapped correctly when expressed in YAML as a string or date.

Let's say I have the following property bean:
```java
@ConfigurationProperties("jcw")
@Data
@Introspected
public class Config {

  LocalDate localDateProperty;
  List<DatesListItem> dates;

  @Builder
  @Data
  @Introspected
  public static class DatesListItem {

    private String id;
    private String stringDate;
    private Date javaUtilDate;
    private LocalDate localDate;
  }

}
```

I'd expect that all examples listed in the `actual behavior` section work properly.

### Actual Behaviour

## Example 1:

```yaml
jcw:
  local-date-property: 2022-01-01
```

**Result**: error `Property doesn't exist` occurs on building context

## Example 2:

```yaml
jcw:
  local-date-property: '2022-01-01'
```

**Result**: error `Unable to convert value [2022-01-01] to target type [LocalDate] due to: Text '2022-01-01' could not be parsed at index 4` occurs on building context

## Example 3:

```yaml
jcw:
  dates:
    - id: stringDate
      stringDate: 2022-01-01
    - id: javaUtilDate
      javaUtilDate: 2022-01-01
    # below entry is not mapped by micronaut because local date is specified as yaml timestamp
    # see https://yaml.org/type/timestamp.html
    - id: localDateAsDate
      localDate: 2022-01-01
    - id: localDateAsString
      localDate: '2022-01-01'
    - id: justId
```
**Result**: element with `localDate: 2022-01-01` is silently discarded




### Steps To Reproduce

see example application with failing tests: https://github.com/juliusz-cwiakalski/micronaut-properties-local-date-issue


```bash
# git clone https://github.com/juliusz-cwiakalski/micronaut-properties-local-date-issue.git
Cloning into 'micronaut-properties-local-date-issue'...
remote: Enumerating objects: 56, done.
remote: Counting objects: 100% (56/56), done.
remote: Compressing objects: 100% (31/31), done.
remote: Total 56 (delta 7), reused 56 (delta 7), pack-reused 0
Receiving objects: 100% (56/56), 57.22 KiB | 1.19 MiB/s, done.
Resolving deltas: 100% (7/7), done.


# cd ./micronaut-properties-local-date-issue


# mvn test
[INFO] Scanning for projects...
[INFO]
[INFO] -------< pl.jcw.bugreport:micronaut-properties-local-date-issue >-------
[INFO] Building micronaut-properties-local-date-issue 0.1
[INFO] --------------------------------[ jar ]---------------------------------
[INFO]
[INFO] --- maven-resources-plugin:3.2.0:resources (default-resources) @ micronaut-properties-local-date-issue ---
[INFO] Using 'UTF-8' encoding to copy filtered resources.
[INFO] Using 'UTF-8' encoding to copy filtered properties files.
[INFO] Copying 5 resources
[INFO]
[INFO] --- maven-compiler-plugin:3.10.1:compile (default-compile) @ micronaut-properties-local-date-issue ---
[INFO] Changes detected - recompiling the module!
[INFO] Compiling 2 source files to /Users/juliuszcwiakalski/git/bugreports/tmp/micronaut-properties-local-date-issue/target/classes
[INFO]
[INFO] --- maven-resources-plugin:3.2.0:testResources (default-testResources) @ micronaut-properties-local-date-issue ---
[INFO] Using 'UTF-8' encoding to copy filtered resources.
[INFO] Using 'UTF-8' encoding to copy filtered properties files.
[INFO] skip non existing resourceDirectory /Users/juliuszcwiakalski/git/bugreports/tmp/micronaut-properties-local-date-issue/src/test/resources
[INFO]
[INFO] --- maven-compiler-plugin:3.10.1:testCompile (default-testCompile) @ micronaut-properties-local-date-issue ---
[INFO] Changes detected - recompiling the module!
[INFO] Compiling 3 source files to /Users/juliuszcwiakalski/git/bugreports/tmp/micronaut-properties-local-date-issue/target/test-classes
[INFO]
[INFO] --- micronaut-maven-plugin:3.4.0:start-testresources-service (default-start-testresources-service) @ micronaut-properties-local-date-issue ---
[INFO]
[INFO] --- maven-surefire-plugin:2.22.2:test (default-test) @ micronaut-properties-local-date-issue ---
[INFO]
[INFO] -------------------------------------------------------
[INFO]  T E S T S
[INFO] -------------------------------------------------------
[INFO] Running pl.jcw.bugreport.LocalDateAsYamlDateTest
10:38:16.676 [main] INFO  i.m.context.env.DefaultEnvironment -
        Established active environments: [test, property-as-date]
[ERROR] Tests run: 1, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 0.692 s <<< FAILURE! - in pl.jcw.bugreport.LocalDateAsYamlDateTest
[ERROR] shouldContainLocalDateExpressedAsStringInYaml  Time elapsed: 0.016 s  <<< ERROR!
io.micronaut.context.exceptions.DependencyInjectionException:
Failed to inject value for parameter [localDateProperty] of method [setLocalDateProperty] of class: pl.jcw.bugreport.Config

Message: Failed to inject value for parameter [localDateProperty] of method [setLocalDateProperty] of class: pl.jcw.bugreport.Config

Message: Error resolving property value [jcw.local-date-property]. Property doesn't exist
Path Taken: LocalDateAsYamlDateTest.config --> Config.setLocalDateProperty([LocalDate localDateProperty])
Path Taken: LocalDateAsYamlDateTest.config --> Config.setLocalDateProperty([LocalDate localDateProperty])

[INFO] Running pl.jcw.bugreport.LocalDateAsYamlStringTest
10:38:17.115 [main] INFO  i.m.context.env.DefaultEnvironment -
        Established active environments: [test, property-as-string]
[ERROR] Tests run: 1, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 0.062 s <<< FAILURE! - in pl.jcw.bugreport.LocalDateAsYamlStringTest
[ERROR] shouldContainLocalDateExpressedAsStringInYaml  Time elapsed: 0.012 s  <<< ERROR!
io.micronaut.context.exceptions.DependencyInjectionException:
Failed to inject value for parameter [localDateProperty] of method [setLocalDateProperty] of class: pl.jcw.bugreport.Config

Message: Failed to inject value for parameter [localDateProperty] of method [setLocalDateProperty] of class: pl.jcw.bugreport.Config

Message: Error resolving property value [jcw.local-date-property]. Unable to convert value [2022-01-01] to target type [LocalDate] due to: Text '2022-01-01' could not be parsed at index 4
Path Taken: LocalDateAsYamlStringTest.config --> Config.setLocalDateProperty([LocalDate localDateProperty])
Path Taken: LocalDateAsYamlStringTest.config --> Config.setLocalDateProperty([LocalDate localDateProperty])

[INFO] Running pl.jcw.bugreport.DatesOnListTest
10:38:17.197 [main] INFO  i.m.context.env.DefaultEnvironment -
        Established active environments: [test, list]
[ERROR] Tests run: 5, Failures: 2, Errors: 0, Skipped: 0, Time elapsed: 0.315 s <<< FAILURE! - in pl.jcw.bugreport.DatesOnListTest
[ERROR] shouldReadFiveEntriesInYaml  Time elapsed: 0.016 s  <<< FAILURE!
java.lang.AssertionError:

Expected size: 5 but was: 4 in:
[Config.DatesListItem(id=stringDate, stringDate=1640995200000, javaUtilDate=null, localDate=null),
    Config.DatesListItem(id=javaUtilDate, stringDate=null, javaUtilDate=Sat Jan 01 01:00:00 CET 2022, localDate=null),
    Config.DatesListItem(id=localDateAsString, stringDate=null, javaUtilDate=null, localDate=2022-01-01),
    Config.DatesListItem(id=justId, stringDate=null, javaUtilDate=null, localDate=null)]
	at pl.jcw.bugreport.DatesOnListTest.shouldReadFiveEntriesInYaml(DatesOnListTest.java:43)

[ERROR] shouldContainLocalDateExpressedAsDateInYaml  Time elapsed: 0.003 s  <<< FAILURE!
java.lang.AssertionError:

Expecting ArrayList:
  [Config.DatesListItem(id=stringDate, stringDate=1640995200000, javaUtilDate=null, localDate=null),
    Config.DatesListItem(id=javaUtilDate, stringDate=null, javaUtilDate=Sat Jan 01 01:00:00 CET 2022, localDate=null),
    Config.DatesListItem(id=localDateAsString, stringDate=null, javaUtilDate=null, localDate=2022-01-01),
    Config.DatesListItem(id=justId, stringDate=null, javaUtilDate=null, localDate=null)]
to contain:
  [Config.DatesListItem(id=localDateAsDate, stringDate=null, javaUtilDate=null, localDate=2022-01-01)]
but could not find the following element(s):
  [Config.DatesListItem(id=localDateAsDate, stringDate=null, javaUtilDate=null, localDate=2022-01-01)]

	at pl.jcw.bugreport.DatesOnListTest.shouldContainLocalDateExpressedAsDateInYaml(DatesOnListTest.java:32)

[INFO]
[INFO] Results:
[INFO]
[ERROR] Failures:
[ERROR]   DatesOnListTest.shouldContainLocalDateExpressedAsDateInYaml:32
Expecting ArrayList:
  [Config.DatesListItem(id=stringDate, stringDate=1640995200000, javaUtilDate=null, localDate=null),
    Config.DatesListItem(id=javaUtilDate, stringDate=null, javaUtilDate=Sat Jan 01 01:00:00 CET 2022, localDate=null),
    Config.DatesListItem(id=localDateAsString, stringDate=null, javaUtilDate=null, localDate=2022-01-01),
    Config.DatesListItem(id=justId, stringDate=null, javaUtilDate=null, localDate=null)]
to contain:
  [Config.DatesListItem(id=localDateAsDate, stringDate=null, javaUtilDate=null, localDate=2022-01-01)]
but could not find the following element(s):
  [Config.DatesListItem(id=localDateAsDate, stringDate=null, javaUtilDate=null, localDate=2022-01-01)]

[ERROR]   DatesOnListTest.shouldReadFiveEntriesInYaml:43
Expected size: 5 but was: 4 in:
[Config.DatesListItem(id=stringDate, stringDate=1640995200000, javaUtilDate=null, localDate=null),
    Config.DatesListItem(id=javaUtilDate, stringDate=null, javaUtilDate=Sat Jan 01 01:00:00 CET 2022, localDate=null),
    Config.DatesListItem(id=localDateAsString, stringDate=null, javaUtilDate=null, localDate=2022-01-01),
    Config.DatesListItem(id=justId, stringDate=null, javaUtilDate=null, localDate=null)]
[ERROR] Errors:
[ERROR]   LocalDateAsYamlDateTest.shouldContainLocalDateExpressedAsStringInYaml » DependencyInjection
[ERROR]   LocalDateAsYamlStringTest.shouldContainLocalDateExpressedAsStringInYaml » DependencyInjection
[INFO]
[ERROR] Tests run: 7, Failures: 2, Errors: 2, Skipped: 0
[INFO]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  5.814 s
[INFO] Finished at: 2022-08-09T10:38:17+02:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-surefire-plugin:2.22.2:test (default-test) on project micronaut-properties-local-date-issue: There are test failures.
[ERROR]
[ERROR] Please refer to /Users/juliuszcwiakalski/git/bugreports/tmp/micronaut-properties-local-date-issue/target/surefire-reports for the individual test results.
[ERROR] Please refer to dump files (if any exist) [date].dump, [date]-jvmRun[N].dump and [date].dumpstream.
[ERROR] -> [Help 1]
[ERROR]
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR]
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoFailureException
```

### Environment Information

- MacOS Monterey 12.5 (21G72)
- java version "17.0.2" 2022-01-18 LTS

### Example Application

https://github.com/juliusz-cwiakalski/micronaut-properties-local-date-issue

### Version

3.6.0

--- a/context/src/main/java/io/micronaut/runtime/converters/time/TimeConverterRegistrar.java
+++ b/context/src/main/java/io/micronaut/runtime/converters/time/TimeConverterRegistrar.java
@@ -41,8 +41,11 @@ import java.time.format.DateTimeFormatter;
 import java.time.format.DateTimeParseException;
 import java.time.temporal.TemporalAccessor;
 import java.time.temporal.TemporalAmount;
+import java.time.temporal.TemporalQuery;
+import java.util.Date;
 import java.util.Optional;
 import java.util.function.BiFunction;
+import java.util.function.Function;
 import java.util.regex.Matcher;
 import java.util.regex.Pattern;
 
@@ -143,23 +146,7 @@ public class TimeConverterRegistrar implements TypeConverterRegistrar {
             (object, targetType, context) -> durationConverter.apply(object, context).map(TemporalAmount.class::cast)
         );
 
-        // CharSequence -> LocalDateTime
-        conversionService.addConverter(
-            CharSequence.class,
-            LocalDateTime.class,
-            (object, targetType, context) -> {
-                try {
-                    DateTimeFormatter formatter = resolveFormatter(context);
-                    LocalDateTime result = LocalDateTime.parse(object, formatter);
-                    return Optional.of(result);
-                } catch (DateTimeParseException e) {
-                    context.reject(object, e);
-                    return Optional.empty();
-                }
-            }
-        );
-
-        // TemporalAccessor - CharSequence
+        // TemporalAccessor -> CharSequence
         final TypeConverter<TemporalAccessor, CharSequence> temporalConverter = (object, targetType, context) -> {
             try {
                 DateTimeFormatter formatter = resolveFormatter(context);
@@ -175,54 +162,58 @@ public class TimeConverterRegistrar implements TypeConverterRegistrar {
                 temporalConverter
         );
 
-        // CharSequence -> LocalDate
-        conversionService.addConverter(
-            CharSequence.class,
-            LocalDate.class,
-            (object, targetType, context) -> {
+        addTemporalStringConverter(conversionService, Instant.class, DateTimeFormatter.ISO_INSTANT, Instant::from);
+        addTemporalStringConverter(conversionService, LocalDate.class, DateTimeFormatter.ISO_LOCAL_DATE, LocalDate::from);
+        addTemporalStringConverter(conversionService, LocalDateTime.class, DateTimeFormatter.ISO_LOCAL_DATE_TIME, LocalDateTime::from);
+        addTemporalStringConverter(conversionService, OffsetTime.class, DateTimeFormatter.ISO_OFFSET_TIME, OffsetTime::from);
+        addTemporalStringConverter(conversionService, OffsetDateTime.class, DateTimeFormatter.ISO_OFFSET_DATE_TIME, OffsetDateTime::from);
+        addTemporalStringConverter(conversionService, ZonedDateTime.class, DateTimeFormatter.ISO_ZONED_DATE_TIME, ZonedDateTime::from);
+
+        // java.time -> Date
+        addTemporalToDateConverter(conversionService, Instant.class, Function.identity());
+        addTemporalToDateConverter(conversionService, OffsetDateTime.class, OffsetDateTime::toInstant);
+        addTemporalToDateConverter(conversionService, ZonedDateTime.class, ZonedDateTime::toInstant);
+        // these two are a bit icky, but required for yaml parsing compatibility
+        addTemporalToDateConverter(conversionService, LocalDate.class, ld -> ld.atTime(0, 0).toInstant(ZoneOffset.UTC));
+        addTemporalToDateConverter(conversionService, LocalDateTime.class, ldt -> ldt.toInstant(ZoneOffset.UTC));
+    }
+
+    private <T extends TemporalAccessor> void addTemporalStringConverter(ConversionService<?> conversionService, Class<T> temporalType, DateTimeFormatter isoFormatter, TemporalQuery<T> query) {
+        conversionService.addConverter(CharSequence.class, temporalType, (CharSequence object, Class<T> targetType, ConversionContext context) -> {
+            if (StringUtils.isEmpty(object)) {
+                return Optional.empty();
+            }
+            // try explicit format first
+            Optional<String> format = context.getAnnotationMetadata().stringValue(Format.class);
+            if (format.isPresent()) {
+                DateTimeFormatter formatter = DateTimeFormatter.ofPattern(format.get(), context.getLocale());
                 try {
-                    DateTimeFormatter formatter = resolveFormatter(context);
-                    LocalDate result = LocalDate.parse(object, formatter);
+                    T converted = formatter.parse(object, query);
+                    return Optional.of(converted);
+                } catch (DateTimeParseException e) {
+                    context.reject(object, e);
+                    return Optional.empty();
+                }
+            } else {
+                try {
+                    T converted = isoFormatter.parse(object, query);
+                    return Optional.of(converted);
+                } catch (DateTimeParseException ignored) {
+                }
+                // fall back to RFC 1123 date time for compatibility
+                try {
+                    T result = DateTimeFormatter.RFC_1123_DATE_TIME.parse(object, query);
                     return Optional.of(result);
                 } catch (DateTimeParseException e) {
                     context.reject(object, e);
                     return Optional.empty();
                 }
             }
-        );
+        });
+    }
 
-
-        // CharSequence -> ZonedDateTime
-        conversionService.addConverter(
-            CharSequence.class,
-            ZonedDateTime.class,
-            (object, targetType, context) -> {
-                try {
-                    DateTimeFormatter formatter = resolveFormatter(context);
-                    ZonedDateTime result = ZonedDateTime.parse(object, formatter);
-                    return Optional.of(result);
-                } catch (DateTimeParseException e) {
-                    context.reject(object, e);
-                    return Optional.empty();
-                }
-            }
-        );
-
-        // CharSequence -> OffsetDateTime
-        conversionService.addConverter(
-                CharSequence.class,
-                OffsetDateTime.class,
-                (object, targetType, context) -> {
-                    try {
-                        DateTimeFormatter formatter = resolveFormatter(context);
-                        OffsetDateTime result = OffsetDateTime.parse(object, formatter);
-                        return Optional.of(result);
-                    } catch (DateTimeParseException e) {
-                        context.reject(object, e);
-                        return Optional.empty();
-                    }
-                }
-        );
+    private <T extends TemporalAccessor> void addTemporalToDateConverter(ConversionService<?> conversionService, Class<T> temporalType, Function<T, Instant> toInstant) {
+        conversionService.addConverter(temporalType, Date.class, (T object, Class<Date> targetType, ConversionContext context) -> Optional.of(Date.from(toInstant.apply(object))));
     }
 
     private DateTimeFormatter resolveFormatter(ConversionContext context) {

--- a/context/src/main/java/io/micronaut/runtime/converters/time/TimeConverterRegistrar.java
+++ b/context/src/main/java/io/micronaut/runtime/converters/time/TimeConverterRegistrar.java
@@ -174,6 +174,7 @@ public class TimeConverterRegistrar implements TypeConverterRegistrar {
         addTemporalToDateConverter(conversionService, OffsetDateTime.class, OffsetDateTime::toInstant);
         addTemporalToDateConverter(conversionService, ZonedDateTime.class, ZonedDateTime::toInstant);
         // these two are a bit icky, but required for yaml parsing compatibility
+        // TODO Micronaut 4 Consider deletion
         addTemporalToDateConverter(conversionService, LocalDate.class, ld -> ld.atTime(0, 0).toInstant(ZoneOffset.UTC));
         addTemporalToDateConverter(conversionService, LocalDateTime.class, ldt -> ldt.toInstant(ZoneOffset.UTC));
     }

--- a/inject/src/main/java/io/micronaut/context/env/yaml/ConstructIsoTimestampString.java
+++ b/inject/src/main/java/io/micronaut/context/env/yaml/ConstructIsoTimestampString.java
@@ -1,0 +1,114 @@
+/*
+ * Copyright 2017-2022 original authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.micronaut.context.env.yaml;
+
+import io.micronaut.core.annotation.Internal;
+import io.micronaut.core.annotation.NonNull;
+import org.yaml.snakeyaml.constructor.AbstractConstruct;
+import org.yaml.snakeyaml.error.YAMLException;
+import org.yaml.snakeyaml.nodes.Node;
+import org.yaml.snakeyaml.nodes.ScalarNode;
+
+import java.time.LocalDate;
+import java.time.LocalDateTime;
+import java.time.ZoneOffset;
+import java.time.temporal.Temporal;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+/**
+ * Adapted from {@link org.yaml.snakeyaml.constructor.SafeConstructor.ConstructYamlTimestamp} but
+ * with java 8 time.
+ */
+@Internal
+final class ConstructIsoTimestampString extends AbstractConstruct {
+    private static final Pattern TIMESTAMP_REGEXP = Pattern.compile(
+        "^([0-9][0-9][0-9][0-9])-([0-9][0-9]?)-([0-9][0-9]?)(?:(?:[Tt]|[ \t]+)([0-9][0-9]?):([0-9][0-9]):([0-9][0-9])(?:\\.([0-9]*))?(?:[ \t]*(?:(Z)|([-+][0-9][0-9]?)(?::([0-9][0-9])?)?))?)?$");
+    private static final Pattern YMD_REGEXP = Pattern
+        .compile("^([0-9][0-9][0-9][0-9])-([0-9][0-9]?)-([0-9][0-9]?)$");
+
+    @Override
+    public Object construct(Node node) {
+        ScalarNode scalar = (ScalarNode) node;
+        String nodeValue = scalar.getValue();
+        return parse(nodeValue);
+    }
+
+    @NonNull
+    static Temporal parse(String nodeValue) {
+        Matcher match = YMD_REGEXP.matcher(nodeValue);
+        if (match.matches()) {
+            String yearS = match.group(1);
+            String monthS = match.group(2);
+            String dayS = match.group(3);
+            return LocalDate.of(
+                Integer.parseInt(yearS),
+                Integer.parseInt(monthS),
+                Integer.parseInt(dayS)
+            );
+        } else {
+            match = TIMESTAMP_REGEXP.matcher(nodeValue);
+            if (!match.matches()) {
+                throw new YAMLException("Unexpected timestamp: " + nodeValue);
+            }
+            String yearS = match.group(1);
+            String monthS = match.group(2);
+            String dayS = match.group(3);
+            String hourS = match.group(4);
+            String minS = match.group(5);
+            // seconds and milliseconds
+            String seconds = match.group(6);
+            String millis = match.group(7);
+            if (millis != null) {
+                seconds = seconds + "." + millis;
+            }
+            double fractions = Double.parseDouble(seconds);
+            int secS = (int) Math.round(Math.floor(fractions));
+            int nsec = (int) Math.round((fractions - secS) * 1_000_000_000);
+
+            LocalDateTime ldt = LocalDateTime.of(
+                Integer.parseInt(yearS),
+                Integer.parseInt(monthS),
+                Integer.parseInt(dayS),
+                Integer.parseInt(hourS),
+                Integer.parseInt(minS),
+                secS,
+                nsec
+            );
+
+            // timezone
+            String timezonehS = match.group(9);
+            String timezonemS = match.group(10);
+            if (timezonehS != null) {
+                ZoneOffset offset;
+                if (timezonemS == null) {
+                    offset = ZoneOffset.ofHours(Integer.parseInt(timezonehS));
+                } else {
+                    offset = ZoneOffset.ofHoursMinutes(Integer.parseInt(timezonehS), Integer.parseInt(timezonemS));
+                }
+                return ldt.atOffset(offset);
+            } else {
+                if (match.group(8) != null) {
+                    // Z
+                    return ldt.atOffset(ZoneOffset.UTC);
+                } else {
+                    // no time zone provided
+                    return ldt;
+                }
+            }
+        }
+    }
+}

--- a/inject/src/main/java/io/micronaut/context/env/yaml/CustomSafeConstructor.java
+++ b/inject/src/main/java/io/micronaut/context/env/yaml/CustomSafeConstructor.java
@@ -19,6 +19,7 @@ import io.micronaut.core.annotation.Internal;
 import org.yaml.snakeyaml.constructor.SafeConstructor;
 import org.yaml.snakeyaml.nodes.MappingNode;
 import org.yaml.snakeyaml.nodes.SequenceNode;
+import org.yaml.snakeyaml.nodes.Tag;
 
 import java.util.List;
 import java.util.Map;
@@ -32,6 +33,9 @@ import java.util.Map;
  */
 @Internal
 class CustomSafeConstructor extends SafeConstructor {
+    CustomSafeConstructor() {
+        yamlConstructors.put(Tag.TIMESTAMP, new ConstructIsoTimestampString());
+    }
 
     @Override
     protected Map<Object, Object> newMap(MappingNode node) {

--- a/inject/src/test/groovy/io/micronaut/context/env/yaml/ConstructIsoTimestampStringSpec.groovy
+++ b/inject/src/test/groovy/io/micronaut/context/env/yaml/ConstructIsoTimestampStringSpec.groovy
@@ -1,0 +1,25 @@
+package io.micronaut.context.env.yaml
+
+import spock.lang.Specification
+
+import java.time.LocalDate
+import java.time.LocalDateTime
+import java.time.ZoneOffset
+import java.time.temporal.Temporal
+
+class ConstructIsoTimestampStringSpec extends Specification {
+    def parse(String literal, Temporal expected) {
+        when:
+        Temporal parsed = ConstructIsoTimestampString.parse(literal)
+        then:
+        parsed == expected
+
+        where:
+        literal                         | expected
+        '2022-08-12'                    | LocalDate.of(2022, 8, 12)
+        '2022-08-12T10:01:02.345'       | LocalDateTime.of(2022, 8, 12, 10, 1, 2, 345_000_000)
+        '2022-08-12T10:01:02.345Z'      | LocalDateTime.of(2022, 8, 12, 10, 1, 2, 345_000_000).atOffset(ZoneOffset.UTC)
+        '2022-08-12T10:01:02.345+05'    | LocalDateTime.of(2022, 8, 12, 10, 1, 2, 345_000_000).atOffset(ZoneOffset.ofHours(5))
+        '2022-08-12T10:01:02.345+05:06' | LocalDateTime.of(2022, 8, 12, 10, 1, 2, 345_000_000).atOffset(ZoneOffset.ofHoursMinutes(5, 6))
+    }
+}

--- a/inject/src/test/groovy/io/micronaut/context/env/yaml/YamlPropertySourceLoaderSpec2.groovy
+++ b/inject/src/test/groovy/io/micronaut/context/env/yaml/YamlPropertySourceLoaderSpec2.groovy
@@ -1,0 +1,93 @@
+/*
+ * Copyright 2017-2019 original authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.micronaut.context.env.yaml
+
+import io.micronaut.context.ApplicationContextConfiguration
+import io.micronaut.context.env.DefaultEnvironment
+import io.micronaut.context.env.Environment
+import io.micronaut.context.env.PropertySourceLoader
+import io.micronaut.core.io.service.ServiceDefinition
+import io.micronaut.core.io.service.SoftServiceLoader
+import spock.lang.Specification
+
+import java.time.LocalDate
+import java.time.LocalDateTime
+import java.time.OffsetDateTime
+import java.time.ZoneOffset
+
+/**
+ * not limited by groovy version
+ */
+class YamlPropertySourceLoaderSpec2 extends Specification {
+
+    void "test yaml value conversion"(String literal, Class<?> type, Object expected) {
+        given:
+        def serviceDefinition = Mock(ServiceDefinition)
+        serviceDefinition.isPresent() >> true
+        serviceDefinition.load() >> new YamlPropertySourceLoader()
+
+        Environment env = new DefaultEnvironment(new ApplicationContextConfiguration() {
+            @Override
+            List<String> getEnvironments() {
+                return ['test']
+            }
+        }) {
+            @Override
+            protected SoftServiceLoader<PropertySourceLoader> readPropertySourceLoaders() {
+                GroovyClassLoader gcl = new GroovyClassLoader()
+                gcl.addClass(YamlPropertySourceLoader)
+                gcl.addURL(YamlPropertySourceLoader.getResource("/META-INF/services/io.micronaut.context.env.PropertySourceLoader"))
+                return new SoftServiceLoader<PropertySourceLoader>(PropertySourceLoader, gcl)
+            }
+
+            @Override
+            Optional<InputStream> getResourceAsStream(String path) {
+                if (path.endsWith("application.yml")) {
+                    return Optional.of(new ByteArrayInputStream("""\
+foo: $literal
+""".bytes))
+                }
+
+                return Optional.empty()
+            }
+
+        }
+
+
+        when:
+        env.start()
+
+        then:
+        env.get("foo", type).get() == expected
+
+        where:
+        // note: string->Date is not supported
+
+        literal                       | type           | expected
+        // YMD
+        '2022-08-12'                  | LocalDate      | LocalDate.of(2022, 8, 12)
+        '"2022-08-12"'                | LocalDate      | LocalDate.of(2022, 8, 12)
+        '2022-08-12'                  | Date           | Date.from(LocalDate.of(2022, 8, 12).atTime(0, 0).atOffset(ZoneOffset.UTC).toInstant())
+        // YMD HMS
+        '2022-08-12T10:12:34'         | LocalDateTime  | LocalDateTime.of(2022, 8, 12, 10, 12, 34)
+        '"2022-08-12T10:12:34"'       | LocalDateTime  | LocalDateTime.of(2022, 8, 12, 10, 12, 34)
+        '2022-08-12T10:12:34'         | Date           | Date.from(LocalDateTime.of(2022, 8, 12, 10, 12, 34).atOffset(ZoneOffset.UTC).toInstant())
+        // YMD HMS Z
+        '2022-08-12T10:12:34+05:00'   | OffsetDateTime | LocalDateTime.of(2022, 8, 12, 10, 12, 34).atOffset(ZoneOffset.ofHours(5))
+        '"2022-08-12T10:12:34+05:00"' | OffsetDateTime | LocalDateTime.of(2022, 8, 12, 10, 12, 34).atOffset(ZoneOffset.ofHours(5))
+        '2022-08-12T10:12:34+05:00'   | Date           | Date.from(LocalDateTime.of(2022, 8, 12, 10, 12, 34).atOffset(ZoneOffset.ofHours(5)).toInstant())
+    }
+}

--- a/runtime/src/test/groovy/io/micronaut/runtime/converters/time/TimeConverterRegistrarSpec.groovy
+++ b/runtime/src/test/groovy/io/micronaut/runtime/converters/time/TimeConverterRegistrarSpec.groovy
@@ -21,6 +21,14 @@ import spock.lang.Specification
 import spock.lang.Unroll
 
 import java.time.Duration
+import java.time.Instant
+import java.time.LocalDate
+import java.time.LocalDateTime
+import java.time.OffsetDateTime
+import java.time.OffsetTime
+import java.time.ZoneId
+import java.time.ZoneOffset
+import java.time.ZonedDateTime
 
 class TimeConverterRegistrarSpec extends Specification {
 
@@ -43,5 +51,30 @@ class TimeConverterRegistrarSpec extends Specification {
         '10h'  | Duration.ofHours(10)
         '10ns' | Duration.ofNanos(10)
 
+    }
+    @Unroll
+    void "test converts a #sourceObject.class.name to a #targetType.name"() {
+        given:
+        ConversionService conversionService = new DefaultConversionService()
+        new TimeConverterRegistrar().register(conversionService)
+
+        expect:
+        conversionService.convert(sourceObject, targetType).get() == result
+
+        where:
+        sourceObject            | targetType  | result
+        Instant.ofEpochMilli(123)                                            | Date | new Date(123)
+        Instant.ofEpochMilli(123).atOffset(ZoneOffset.ofHours(5))            | Date | new Date(123)
+        Instant.ofEpochMilli(123).atZone(ZoneId.of("Europe/Berlin"))         | Date | new Date(123)
+        Instant.ofEpochMilli(123).atOffset(ZoneOffset.UTC).toLocalDateTime() | Date | new Date(123)
+        Instant.ofEpochMilli(123).atOffset(ZoneOffset.UTC).toLocalDate()     | Date | new Date(0)
+
+        "2022-08-12"                               | LocalDate      | LocalDate.of(2022, 8, 12)
+        "2022-08-12T12:19:00"                      | LocalDateTime  | LocalDateTime.of(2022, 8, 12, 12, 19)
+        "12:19:00+05:00"                           | OffsetTime     | OffsetTime.of(12, 19, 0, 0, ZoneOffset.ofHours(5))
+        "2022-08-12T12:19:00+05:00"                | OffsetDateTime | OffsetDateTime.of(2022, 8, 12, 12, 19, 0, 0, ZoneOffset.ofHours(5))
+        "2022-08-12T12:19:00+05:00"                | ZonedDateTime  | ZonedDateTime.of(2022, 8, 12, 12, 19, 0, 0, ZoneOffset.ofHours(5))
+        "2022-08-12T12:19:00+02:00[Europe/Berlin]" | ZonedDateTime  | ZonedDateTime.of(2022, 8, 12, 12, 19, 0, 0, ZoneId.of("Europe/Berlin"))
+        "2022-08-12T12:19:00Z"                     | Instant        | LocalDateTime.of(2022, 8, 12, 12, 19).toInstant(ZoneOffset.UTC)
     }
 }

